src/io/simulator/simulator_handle.hpp

```diff
@@ -0,0 +1,453 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <compare>
+
+#include <any>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <optional>
+#include <set>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include "io/address.hpp"
+#include "io/errors.hpp"
+#include "io/simulator/simulator_config.hpp"
+#include "io/simulator/simulator_stats.hpp"
+#include "io/transport.hpp"
+
+namespace memgraph::io::simulator {
+struct OpaqueMessage {
+  Address from_address;
+  uint64_t request_id;
+  std::any message;
+
+  /// Recursively tries to match a specific type from the outer
+  /// variant's parameter pack against the type of the std::any,
+  /// and if it matches, make it concrete and return it. Otherwise,
+  /// move on and compare the any with the next type from the
+  /// parameter pack.
+  ///
+  /// Return is the full std::variant<Ts...> type that holds the
+  /// full parameter pack without interfering with recursive
+  /// narrowing expansion.
+  template <typename Return, Message Head, Message... Rest>
+  std::optional<Return> Unpack(std::any &&a) {
+    if (typeid(Head) == a.type()) {
+      Head concrete = std::any_cast<Head>(std::move(a));
+      return concrete;
+    }
+
+    if constexpr (sizeof...(Rest) > 0) {
+      return Unpack<Return, Rest...>(std::move(a));
+    } else {
+      return std::nullopt;
+    }
+  }
+
+  /// High level "user-facing" conversion function that lets
+  /// people interested in conversion only supply a single
+  /// parameter pack for the types that they want to compare
+  /// with the any and potentially include in the returned
+  /// variant.
+  template <Message... Ms>
+  requires(sizeof...(Ms) > 0) std::optional<std::variant<Ms...>> VariantFromAny(std::any &&a) {
+    return Unpack<std::variant<Ms...>, Ms...>(std::move(a));
+  }
+
+  template <Message... Ms>
+  requires(sizeof...(Ms) > 0) std::optional<RequestEnvelope<Ms...>> Take() {
+    std::optional<std::variant<Ms...>> m_opt = VariantFromAny<Ms...>(std::move(message));
+
+    if (m_opt) {
+      return RequestEnvelope<Ms...>{
+          .message = std::move(*m_opt),
+          .request_id = request_id,
+          .from_address = from_address,
+      };
+    } else {
+      return std::nullopt;
+    }
+  }
+};
+
+struct PromiseKey {
+  Address requester_address;
+  uint64_t request_id;
+  Address replier_address;
+
+ public:
+  bool operator<(const PromiseKey &other) const {
+    if (requester_address == other.requester_address) {
+      return request_id < other.request_id;
+    } else {
+      return requester_address < other.requester_address;
+    }
+  }
+};
+
+class OpaquePromise {
+  const std::type_info *ti_;
+  void *ptr_;
+  std::function<void(void *)> dtor_;
+  std::function<bool(void *)> is_awaited_;
+  std::function<void(void *, OpaqueMessage)> fill_;
+  std::function<void(void *)> time_out_;
```

Comment:
I think 4 `std::function` is a good indicator to introduce virtual inheritance :smiley:
Something like this should work:
```
class OpaquePromiseTraitBase {
    virtual const std::type_info * TypeInfo() const = 0;
    virtual void Destroy(void* ptr) const = 0;
    virtual bool IsAwaited(void* ptr) const = 0;
    virtual bool Fill(void* ptr, OpaqueMessage) const = 0;
    virtual void Timeout() const = 0;
};

template <typename T>
class OpaquePromieTrait : public OpaquePromiseTraitBase {
     const std::type_info * TypeInfo() const override { return &typeid(T);};

     void Destroy(void* ptr) const override {
        // I think you need to call delete on the pointer,, as the destructor won't free the memory.
        delete static_cast<ResponsePromise<T> *>(ptr);
     };

     bool IsAwaited(void* ptr) const override {
        return static_cast<ResponsePromise<T> *>(ptr)->IsAwaited(); 
     };
     
     bool Fill(void* ptr, OpaqueMessage msg) const override {
        // ...
     };
     
     void Timeout() const override {
        // ...
     }
};


template <typename T>
class OpaquePromise {
  // Of course, ctor is necessary but left it out now
  void *ptr_;
  std::unique_ptr<OpaquePromiseTraitBase> trait_{std::make_unique<OpaquePromieTrait<T>>()};
};
```

Reply:
As a bonus, you can also default the move operators.

Reply:
You also missed this in my opinion.

Reply:
.

Reply:
applied, thank you